internal/pkg/levee/propagation/propagation.go

```diff
@@ -243,6 +246,45 @@ func (prop *Propagation) visitOperands(n ssa.Node, maxInstrReached map[*ssa.Basi
 	}
 }
 
+func (prop *Propagation) visitSelect(sel *ssa.Select, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	// The Send operations have to be processed before the Recv operations in case a channel
+	// is written to in one branch and read from in a different branch. This matters if the
+	// branches are reachable from each other, e.g. when the select is in a loop.
+	for _, s := range sel.States {
+		switch s.Dir {
+		case types.SendOnly:
+			if prop.marked[s.Send.(ssa.Node)] {
+				prop.dfs(s.Chan.(ssa.Node), maxInstrReached, lastBlockVisited, false)
+			}
+		}
+	}
+
+	if sel.Referrers() == nil {
+		return
+	}
+
+	recvStateCount := 0
+	for _, s := range sel.States {
+		switch s.Dir {
+		case types.RecvOnly:
```

Comment:
This could also use a top-level clarifying comment.  `// Propagate taint when receiving a from a tainted channel."

-----

The separation of \<examine all send\> and \<examine all recv\> makes me uncomfortable.  This feels more like a BFS spread out than two disjoint considerations.

-----

I was toying with daisy-chain `select` blocks to explore that separation of Send and Recv logic.  I note that...

```golang
func TestFoo(srcs chan core.Source, i1, i2, i3, i4 chan interface{}) {
	select {
	case x := <-i3:
		i4 <- x
	case y := <-i2:
		i3 <- y
	case z := <-i1:
		i2 <- z
	case s := <-srcs:
		i1 <- s
	}
	core.Sink(srcs) // want "a source has reached a sink"
	core.Sink(i1)   // want "a source has reached a sink"
	core.Sink(i2)   // TODO want "a source has reached a sink"
	core.Sink(i3)   // TODO want "a source has reached a sink"
	core.Sink(i4)   // want "a source has reached a sink"
}

func TestBar(srcs chan core.Source, i1, i2, i3, i4 chan interface{}) {
	select {
	case s := <-srcs:
		i1 <- s
	case z := <-i1:
		i2 <- z
	case y := <-i2:
		i3 <- y
	case x := <-i3:
		i4 <- x
	}
	core.Sink(srcs) // want "a source has reached a sink"
	core.Sink(i1)   // want "a source has reached a sink"
	core.Sink(i2)   // TODO want "a source has reached a sink"
	core.Sink(i3)   // TODO want "a source has reached a sink"
	core.Sink(i4)   // TODO want "a source has reached a sink"
}
```

The order-dependent tainting of `i4` is concerning.

Reply:
Woops. Good old "increment at the end of loop body isn't reached because of earlier continue so when the channel is actually tainted the wrong extract is visited".

Comments + suggested tests added. I've also added a relevant issue number (#211) to the `TODO`s.